### ISSUE TITLE
[competitor-port] Strategy framework & hyperparameter optimization: unified strategy interface + Optuna-based hyperopt + informative pairs (freqtrade + nautilus patterns)

### ISSUE THREAD

**Issue by KCNyu** [KCNyu/clawock#1154]
## 竞品做法
**freqtrade** — `freqtrade/strategy/interface.py` + `freqtrade/optimize/hyperopt/`（源码：https://github.com/freqtrade/freqtrade/tree/develop/freqtrade/strategy、https://github.com/freqtrade/freqtrade/tree/develop/freqtrade/optimize/hyperopt）

- **IStrategy 抽象基类**：定义标准化策略接口
  - `populate_indicators()`：计算技术指标（返回 DataFrame）
  - `populate_entry_trend()` / `populate_exit_trend()`：生成进场/离场信号
  - `populate_informative_pair()`：`@informative` 装饰器支持多时间框架/多交易对特征工程
  - 标准化属性：`minimal_roi`、`stoploss`、`trailing_stop`、`timeframe`、`max_open_trades`、`order_types`、`can_short`、`protections` 等
  - 生命周期钩子：`bot_start()`、`bot_loop_start()`、`check_buy_timeout()`、`check_sell_timeout()`、`adjust_stop_loss()`、`adjust_position()`、`custom_stake_amount()`、`leverage()` 等
  - `use_exit_signal`、`ignore_roi_if_entry_signal`、`process_only_new_candles` 等开关

- **Hyperopt 系统**：基于 Optuna 的贝叶斯优化框架
  - `Hyperopt` 类：并行优化引擎，支持 `Parallel(n_jobs)` 多进程加速
  - `HyperOptimizer`：Optuna 采样器封装，支持 TPE/CMA-ES/NSGA-II 等采样器
  - `IHyperOptLoss` 接口：`hyperopt_loss_function(results, trade_count, min_date, max_date, config, processed, backtest_stats, starting_balance)` 返回标量损失
  - 内置损失函数族：`ShortTradeDurHyperOptLoss`、`SortinoHyperOptLoss`、`CalmarHyperOptLoss`、`MaxDrawDownHyperOptLoss`、`OnlyProfitHyperOptLoss`、`ProfitDrawDownHyperOptLoss`、`SharpeHyperOptLoss`、`MultiMetricHyperOptLoss` 等
  - `HyperoptLossInterface` 基类，`hyperopt_loss_function()` 返回标量损失
  - 结果持久化：`.fthypt` 文件（JSONL 格式），支持断点续跑、早停（ES）
  - 参数去重：`duplicate_optuna_asked_points()` 防止重复评估
  - 早停：`es_epochs` + `es_terminator` 基于 Optuna 的终止器

- **Informative Pairs / Multi-timeframe**：`@informative` 装饰器自动合并不同时间框架数据，支持跨交易对特征工程

**nautilus_trader** — `crates/trading/src/strategy/`（源码：https://github.com/nautechsystems/nautilus_trader/tree/develop/crates/trading/src/strategy）

- **Strategy Trait**：`trait Strategy: DataActor` 核心特征
  - 继承 `DataActor`（数据订阅、事件处理、定时器）
  - `submit_order()` / `submit_order_list()` / `modify_order()` / `cancel_order()` 统一订单 API
  - `order()` / `portfolio()` / `cache()` / `clock()` 门面方法
  - `on_order_*` / `on_trade_*` / `on_position_*` 事件回调
  - `nautilus_strategy!` 宏自动生成运行时 wiring
  - `StrategyConfig` / `ImportableStrategyConfig` 配置序列化
  - `StrategyId` / `TraderId` / `AccountId` 标识符体系

## 我们现在怎么做的
`src/clawock/evaluation/add_alpha_walkforward.py` — 单一硬编码评估流程，**无策略抽象、无参数优化、无多时间框架支持**
- `add_alpha.confirmation_levels()` 等硬编码逻辑，**无策略接口**
- `_confirmation_fill()` 等硬编码成交模拟，**无可插拔执行模型**
- `_cluster_ci()` 等统计工具，**无参数优化框架**
- `add_alpha_walkforward.py:382-629` — `evaluate()` 单一评估函数，**无策略抽象层**
- `src/clawock/decision/add_alpha.py` — `confirmation_levels()` 等确认逻辑，**无策略接口**
- `src/clawock/decision/ledger.py` — 决策账本，**无策略抽象、无优化框架**

## 差在哪
**机制缺失**：
1. **无统一策略接口** — 现有逻辑硬编码在 `add_alpha_walkforward.py`、`add_alpha.py` 中，无法热插拔、无法复用、无法测试
2. **无参数优化框架** — 无 Optuna/贝叶斯优化集成，参数手工调优，无法自动搜索最优超参数
3. **无多时间框架/多交易对特征工程** — 无 Informative Pairs 机制，无法利用高频/跨品种信息
3. **无策略生命周期管理** — 无 `bot_start`/`bot_loop_start` 等生命周期钩子，无法管理策略状态
4. **无标准化损失函数** — 无 `IHyperOptLoss` 接口，无法插拔不同优化目标（Sharpe/Calmar/MaxDD/Sortino/自定义）
5. **无并行优化/早停** — 无 joblib/Optuna 并行加速，无早停机制，无断点续跑

## 改法
新建/改动文件（≤400 LOC，纯离线，无新增外部服务、无新增 cron）：

1. **新建 `src/clawock/strategy/interface.py`**（~150 LOC）
   - `IStrategy` 抽象基类：`populate_indicators()`、`populate_entry_trend()`、`populate_exit_trend()`、`populate_informative()`、`bot_start()`、`bot_loop_start()`、`check_buy_timeout()`、`check_sell_timeout()`、`adjust_stop_loss()`、`adjust_position()`、`custom_stake_amount()`、`leverage()`、`confirmation_levels()` 等标准接口
   - 标准属性：`minimal_roi`、`stoploss`、`trailing_stop`、`timeframe`、`max_open_trades`、`order_types`、`can_short`、`protections`、`use_exit_signal`、`exit_profit_only`、`ignore_roi_if_entry_signal`、`process_only_new_candles`、`startup_candle_count`、`plot_config`
   - `IStrategyConfig` Pydantic 模型：序列化/反序列化配置
   - `InformativePair` 数据类：`pair`、`timeframe`、`candle_type`、`indicators` 列表

2. **新建 `src/clawock/strategy/hyperopt.py`**（~150 LOC）
   - `HyperOptimizer`：Optuna 采样器封装（TPE/CMA-ES/NSGA-II），`study.optimize()` 封装
   - `HyperoptLoss` 抽象基类：`hyperopt_loss_function(results, trade_count, min_date, max_date, config, processed, backtest_stats, starting_balance) -> float`
   - 内置损失函数：`SharpeLoss`、`SortinoLoss`、`CalmarLoss`、`MaxDrawdownLoss`、`MultiMetricLoss`、`ProfitFactorLoss`
   - `HyperoptRunner`：`run_optimization(config, strategy_class, data_provider, n_trials, n_jobs)` 并行优化入口
   - `HyperoptStateContainer`：状态持久化（`.fthypt` JSONL 文件）、断点续跑、早停（ES）

3. **新建 `src/clawock/strategy/informative.py`**（~80 LOC）
   - `InformativePair` / `InformativeCache`：多时间框架数据合并
   - `populate_informative()` 标准接口，`@informative` 装饰器语法糖
   - `merge_informative_pair()`：自动对齐时间索引、前向填充缺失值

4. **改 `src/clawock/evaluation/add_alpha_walkforward.py`**（~60 LOC 增量）
   - 将硬编码逻辑重构为 `AlphaStrategy(IStrategy)` 实现
   - `main()` 新增 `--strategy`/`--hyperopt`/`--n-trials`/`--loss` 参数
   - 集成 `HyperoptRunner`，支持 `--hyperopt` 模式自动优化

**规模量级**：~400 LOC 新增/改动，纯 Python，依赖 `optuna`、`joblib`（已在 `pyproject.toml`）、现有 `pandas`/`numpy`。

## 验收
今天会失败、改完会通过的 pytest 断言（贴在测试文件 `tests/test_strategy_framework.py`）：

```python
def test_strategy_interface_enforces_interface():
    from clawock.strategy.interface import IStrategy
    import pytest
    class BadStrategy(IStrategy):
        pass
    with pytest.raises(TypeError):
        BadStrategy(config={})  # 缺少抽象方法实现

def test_strategy_informative_pair_merging():
    from clawock.strategy.informative import InformativePair, merge_informative_pair
    import pandas as pd
    import numpy as np
    base = pd.DataFrame({"close": [100, 101, 102]}, index=pd.date_range("2025-01-01", periods=3, freq="D"))
    inf = pd.DataFrame({"rsi": [60, 65, 70]}, index=pd.date_range("2025-01-01", periods=3, freq="D"))
    merged = merge_informative_pair(base, inf, "rsi", "1h")
    assert "rsi" in merged.columns
    assert merged["rsi"].iloc[-1] == 70

def test_hyperopt_loss_sharpe():
    from clawock.strategy.hyperopt import SharpeLoss
    import pandas as pd
    import numpy as np
    np.random.seed(42)
    returns = pd.Series(np.random.randn(100) * 0.01)
    backtest_stats = {"sharpe_ratio": 1.5, "max_drawdown": -0.1, "total_trades": 50}
    loss = SharpeLoss().hyperopt_loss_function(
        results=returns.to_frame("returns"),
        trade_count=50,
        min_date="2025-01-01",
        max_date="2025-04-10",
        config={},
        processed={},
        backtest_stats=backtest_stats,
        starting_balance=10000,
    )
    assert isinstance(loss, float)
    assert loss < 0  # 负 Sharpe 作为损失

def test_hyperopt_runner_parallel():
    from clawock.strategy.hyperopt import HyperoptRunner
    import pytest
    class DummyStrategy:
        def populate_indicators(self, df, meta): return df
        def populate_entry_trend(self, df, meta): return df
        def populate_exit_trend(self, df, meta): return df
    config = {"n_trials": 5, "n_jobs": 2}
    runner = HyperoptRunner(config)
    # 仅测试 runner 初始化不报错
    assert runner.config["n_trials"] == 5

def test_informative_pair_merging():
    from clawock.strategy.informative import merge_informative_pair, InformativePair
    import pandas as pd
    base = pd.DataFrame({"close": [100, 101, 102]}, index=pd.date_range("2025-01-01", periods=3, freq="D"))
    inf = InformativePair(pair="BTCUSDT", timeframe="1h", data=pd.DataFrame({"rsi": [60, 65, 70]}, index=pd.date_range("2025-01-01", periods=3, freq="H")))
    merged = merge_informative_pair(base, inf, "rsi", "1h")
    assert "rsi" in merged.columns
```

## ROI 与风险
- **影响面**：`add_alpha_walkforward` 重构为可插拔策略，支持自动超参数优化，`daily-deep-brief` 技能可复用策略接口
- **高压线检查**：
  - live 每日流水线模型输出契约：**不碰** — 只在离线评估/优化脚本里引入策略框架
  - 预注册研究契约 (`peer_residuals.load_rule_config`)：**不碰**
  - dashboard 200KB payload 闸：**不碰**
- **风险**：Optuna 依赖增加约 50MB；搜索空间过大时优化耗时长，需配置 `n_trials`/`timeout` 兜底；多时间框架数据对齐需处理时区/缺失值边界情况

**Comment by KCNyu** [comment]
## 关闭:这条要装的是一台搜索引擎,而本仓库最主要的防线就是不搜索

`#1154` 要 **Optuna-based hyperopt**:给策略参数做超参搜索。

这与本仓库的核心研究纪律**正面冲突**,而且不是风格分歧,是同一件事的两面:

- `add_alpha_walkforward` 的 run card 里写着 `"parameter_fit": "none; pre-registered thresholds"`;
- `evaluation/cscv.py` 存在的唯一理由是**给「搜索过多少个配置」定价**(PBO);
- `evaluation/deflated_sharpe.py` 存在的唯一理由是**把观测 Sharpe 减去「搜索 N 次的期望最大值」**;
- `peer_residuals.load_rule_config` 硬要求预注册,`#1122` 之所以卡着,就是因为换宇宙等于重注册;
- `evaluation/unified.py` 明写 **`validated` 这个词永远给不出来** —— 它属于「在数据到来之前就被写下来」的规则。

**在这样一套东西旁边加一个超参优化器,等于给一个主要防御是「不搜索」的系统装上搜索引擎。**
`#1149` 那条统一评估里量到的东西正好说明后果:同样长度的纯噪声,四个配置里挑最好的那个,
PBO 0.80、DSR 0.68 —— 而 Optuna 一轮就是几百个配置。

「用 CPCV 把 hyperopt 包起来就安全了」也不成立:CPCV **量**过拟合,不**消除**它。
本书的评估样本是 **18 个注册 session**(截面 13 只票)。在这个样本上跑超参搜索,
唯一诚实的产出是一个很大的 PBO 数字。

`#1154` 里**真正可移植且已经做掉**的那一半是「统一策略接口」:
`decision/add_alpha.py` + `POLICY_VERSION` + `config/add-alpha-policy.json` 已经是这个形状 ——
参数在配置里、版本化、被 run card 记录(`#1139` 之后连配置文件内容哈希和随机种子都进 key 了)。
缺的从来不是接口,是「可以搜索」这个前提。

将来样本足够到可以承担一次搜索时,重开一条**带样本量论证 + 预注册搜索空间 + 事前声明的 PBO 阈值**的
issue,比复活这条更省事。
